frb_rust/src/ffi.rs

```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    impl<T: IntoDart + Clone> IntoDart for &T {
+        #[inline]
+        fn into_dart(self) -> DartCObject {
+            self.clone().into_dart()
+        }
+    }
+
+    #[derive(Clone)]
+    pub struct Isolate {
+        port: MessagePort,
+    }
+
+    impl Isolate {
+        pub fn new(port: MessagePort) -> Self {
+            Self { port }
+        }
+        pub fn post(&self, msg: impl IntoDart) -> bool {
+            let err = self.port.post_message(&msg.into_dart());
+            if let Err(err) = &err {
+                #[cfg(target_family = "wasm")]
+                crate::console_error!("post: {:?}", err);
+            }
+            err.is_ok()
+        }
+    }
+
+    #[wasm_bindgen]
+    extern "C" {
+        #[wasm_bindgen(js_namespace = console)]
+        pub fn log(msg: &str);
+        #[wasm_bindgen(js_namespace = console)]
+        pub fn error(msg: &str);
+        // fn eval(script: &str);
+    }
+
+    type RawClosure<T> = Box<dyn FnOnce(&[T]) + Send + 'static>;
+
+    pub struct TransferClosure<'a, T> {
+        transfer: &'a [T],
+        closure: RawClosure<T>,
+    }
+
+    pub struct TransferClosurePayload {
+        func: RawClosure<JsValue>,
+    }
+
+    impl<'a> TransferClosure<'a, JsValue> {
+        pub fn new(
+            transfer: &'a [JsValue],
+            closure: impl FnOnce(&[JsValue]) + Send + 'static,
+        ) -> Self {
+            Self {
+                transfer,
+                closure: Box::new(closure),
+            }
+        }
+        /// Posts a <code>[*mut [TransferClosurePayload], ...[JsValue]]</code> message to this worker.
+        ///
+        /// The worker's onmessage should run the corresponding [receive_transfer_closure]
+        /// to receive the message.
+        pub fn apply(self, worker: &Worker) -> Result<(), JsValue> {
+            let transfer = Array::from_iter(self.transfer);
+            // The worker is responsible for cleaning up the leak here.
+            let payload = Box::leak(Box::new(TransferClosurePayload { func: self.closure }));
+            let data = Array::from(&transfer);
+            data.unshift(&JsValue::from(payload as *mut _ as u32));
+            worker
+                .post_message_with_transfer(&data, &transfer)
+                .map_err(|err| {
+                    // post message failed, ownership remains with us.
+                    drop(unsafe { Box::from_raw(payload) });
+                    err
+                })
+        }
+    }
+
+    #[wasm_bindgen]
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
+    pub fn receive_transfer_closure(
+        payload: *mut TransferClosurePayload,
+        transfer: Box<[JsValue]>,
+    ) -> Result<(), JsValue> {
+        let TransferClosurePayload { func } = unsafe { *Box::from_raw(payload) };
+        func(&transfer);
+        global()
+            .unchecked_into::<DedicatedWorkerGlobalScope>()
+            .post_message(&JsValue::undefined())
+    }
```

Comment:
As there are a few non-trivial unsafe operations, we should be serious about testing memory problems for this web implementation...

P.S. Surely no need to implement now, those tests can be deferred to much later :)